cpp/serve/radix_tree.h

```diff
@@ -0,0 +1,398 @@
+
+  /*!
+   * \brief Add an empty sequence at root.
+   * \param seq_id The new sequence ID.
+   * \throw Error if sequence ID is not valid.
+   */
+  void AddSequence(int64_t seq_id);
+
+  /*!
+   * \brief Extend a sequence with given tokens.
+   * \param seq_id The sequence ID for index.
+   * \param tokens The given tokens to extend.
+   * \throw Error if sequence ID is not valid.
+   */
+  void ExtendSequence(int64_t seq_id, IntTuple tokens);
+
+  /*!
+   * \brief Remove a sequence.
+   * \param seq_id The sequence ID to remove.
+   * \throw Error if sequence ID is not valid.
+   */
+  void RemoveSequence(int64_t seq_id);
+
+  /*!
+   * \brief Get the remaining token capacity of the paged radix tree.
+   * \return The the remaining token capacity of the paged radix tree.
+   */
+  size_t FreeCapacity();
+
+  /*! \brief The destructor to free root page. */
+  ~PagedRadixTreeObj();
+
+  static constexpr const uint32_t _type_index = TypeIndex::kDynamic;
+  static constexpr const char* _type_key = "mlc.serve.PagedRadixTree";
+  TVM_DECLARE_FINAL_OBJECT_INFO(PagedRadixTreeObj, Object)
+
+ private:
+  /*!
+   * \brief Merge a radix tree page with its child radix tree page, to save radix tree page.
+   * e.g. MergePage([1, 2, _, _, _] -> [3, 4, 5, _, _]) = [1, 2, 3, 4, 5].
+   * And the page to be merged should be page->Mergeable().
+   * \param page The parent radix tree page.
+   */
+  void MergePage(RedixPage* page);
+
+  /*!
+   * \brief Split a radix tree page at given postition, to accept new sequence.
+   * e.g. SplitPage([1, 2, 3, 4, 5], 2) = [1, 2, _, _, _] -> [3, 4, 5, _, _].
+   * \param page The radix tree page to split.
+   * \param offset The position to split the radix tree page.
+   * \return The splitted radix tree page. It can be different from the input radix tree page, as
+   * there may be implicit radix tree page merge.
+   */
+  RedixPage* SplitPage(RedixPage* page, size_t offset);
+
+  /*!
+   * \brief Match with given token from a radix tree page, stopping at first mismatch.
+   * \param page The radix tree page to start matching.
+   * \param tokens The given tokens to match.
+   * \param length The length of given tokens.
+   */
+  std::tuple<RedixPage*, size_t, size_t> MatchSequence(RedixPage* page, const int64_t* tokens,
+                                                       size_t length);
+};
+
+TVM_REGISTER_OBJECT_TYPE(PagedRadixTreeObj);
+
+class PagedRadixTree : public ObjectRef {
+ public:
+  /*!
+   * \brief Constructor of paged radix tree.
+   * \param num_pages The number of radix tree pages.
+   * \param page_size The page size of each radix tree page.
+   * \param num_seqs The maximum number of sequence ID.
+   */
+  PagedRadixTree(size_t num_pages, size_t page_size, size_t num_seqs);
+
+  TVM_DEFINE_MUTABLE_OBJECT_REF_METHODS(PagedRadixTree, ObjectRef, PagedRadixTreeObj);
+};
+}  // namespace serve
+}  // namespace llm
+}  // namespace mlc
+
+#endif  // MLC_LLM_SERVE_RADIX_TREE_H_
```

Comment:
Note the newline (some editor like VSCode can automatically add newline after the file)